compiler/codegen/OMRCodeGenerator.cpp

```diff
@@ -2059,36 +2059,31 @@ OMR::CodeGenerator::compute64BitMagicValues(
 
    // Cache some common denominators and their magic values.  The key values in this
    // array MUST be in numerically increasing order for the binary search to work.
-   //
-   // The table is composed of 32-bit values because the compiler seems to have a problem
-   // statically initializing it with int64_t constant values.
 
    #define NUM_64BIT_MAGIC_VALUES 6
-   #define TOINT64(x) (*( (int64_t *) &x))
-   static uint32_t div64BitMagicValues[NUM_64BIT_MAGIC_VALUES][6] =
-
-   //     Denominator        Magic Value          Shift
+   static uint64_t div64BitMagicValues[NUM_64BIT_MAGIC_VALUES][3] =
+   //     Denominator                     Magic Value   Shift
 
-      { {    3, 0,    0x55555556, 0x55555555,    0, 0 },
-        {    5, 0,    0x66666667, 0x66666666,    1, 0 },
-        {    7, 0,    0x24924925, 0x49249249,    1, 0 },
-        {    9, 0,    0x71c71c72, 0x1c71c71c,    0, 0 },
-        {   10, 0,    0x66666667, 0x66666666,    2, 0 },
-        {   12, 0,    0xaaaaaaab, 0x2aaaaaaa,    1, 0 } };
+      { {           3, CONSTANT64(0x5555555555555556),      0 },
+        {           5, CONSTANT64(0x6666666666666667),      1 },
+        {           7, CONSTANT64(0x4924924924924925),      1 },
+        {           9, CONSTANT64(0x1c71c71c71c71c72),      0 },
+        {          10, CONSTANT64(0x6666666666666667),      2 },
+        {          12, CONSTANT64(0x2aaaaaaaaaaaaaab),      1 } };
 
    // Quick check if 'd' is cached.
    first = 0;
    last  = NUM_64BIT_MAGIC_VALUES-1;
    while (first <= last)
       {
       mid = (first + last) / 2;
-      if (TOINT64(div64BitMagicValues[mid][0]) == d)
+      if (div64BitMagicValues[mid][0] == d)
          {
-         *m = TOINT64(div64BitMagicValues[mid][2]);
-         *s = TOINT64(div64BitMagicValues[mid][4]);
+         *m = div64BitMagicValues[mid][1];
+         *s = div64BitMagicValues[mid][2];
          return;
          }
-      else if (d > TOINT64(div64BitMagicValues[mid][0]))
+      else if (d > div64BitMagicValues[mid][0])
```

Comment:
I think you still need the `int64_t` conversion when extracting the divisor from this table of UNSIGNED constants to ensure the binary search operates the way it's intended to.   This is because these magic values are to be used for SIGNED division and hence the divisors (and the divisor column) are signed quantities.

I might be mistaken with the C++ implicit type conversion rules (ahem @Leonardo2718), but I believe this comparison will be done as an unsigned comparison.  This means that if this table is ever expanded to include negative divisors then it may not work as expected.

This applies to the other use above as well.

Reply:
Could we make it a table of  SIGNED constants?

Reply:
Since they're initialized with hex constants then I'd say yes.   In fact, the API to this function returns the table values as signed quantities anyway.  However, I will caution the author to do due diligence on this just to be sure it's a correct thing to do.

Reply:
I've taken a closer look and I don't see anything that relies on the fact that the values are unsigned. In fact, all uses of these values besides the comparisons during the binary search seem to immediately convert the values to signed integers and the constants being loaded in are actually signed integers being converted to unsigned integers.

On that note, `compute32BitMagicValues` seems to have the same issue, so I'll go ahead and fix that while I'm at it.